src/Microsoft.Tye.Hosting/ReplicaMonitor.cs

```diff
@@ -0,0 +1,448 @@
+            switch (replicaEvent.State)
+            {
+                case ReplicaState.Started:
+                    _states.TryAdd(replicaEvent.Replica.Name, new ReplicaMonitorState(replicaEvent.Replica, _logger));
+                    break;
+                case ReplicaState.Stopped:
+                    if (_states.TryRemove(replicaEvent.Replica.Name, out var stateToDispose))
+                    {
+                        stateToDispose.Dispose();
+                    }
+                    break;
+                default:
+                    if (_states.TryGetValue(replicaEvent.Replica.Name, out var state))
+                    {
+                        state.Update(replicaEvent);
+                    }
+                    break;
+            }
+        }
+
+        private class Subscription
+        {
+        }
+
+        private class ReplicaMonitorState : IDisposable
+        {
+            private ReplicaStatus _replica;
+            private ILogger _logger;
+
+            private Prober? _livenessProber;
+            private Prober? _readinessProber;
+            private IDisposable? _livenessProberObserver;
+            private IDisposable? _readinessProberObserver;
+
+            private ReplicaState _currentState;
+            private DateTime _lastStateChange;
+            private object _stateChangeLocker;
+
+            public ReplicaMonitorState(ReplicaStatus replica, ILogger logger)
+            {
+                _replica = replica;
+                _logger = logger;
+
+                _currentState = ReplicaState.Started;
+                _lastStateChange = DateTime.Now;
+                _stateChangeLocker = new object();
+
+                Init();
+            }
+
+            private void Init()
+            {
+                var serviceDesc = _replica.Service.Description;
+                if (serviceDesc.Liveness is null && serviceDesc.Readiness is null)
+                {
+                    MoveToReady();
+                }
+                else if (serviceDesc.Liveness is null)
+                {
+                    MoveToHealthy(from: ReplicaState.Started);
+                    StartReadinessProbe(serviceDesc.Readiness!);
+                }
+                else if (serviceDesc.Readiness is null)
+                {
+                    StartLivenessProbe(serviceDesc.Liveness, moveToOnSuccess: ReplicaState.Ready);
+                }
+                else
+                {
+                    StartLivenessProbe(serviceDesc.Liveness);
+                    StartReadinessProbe(serviceDesc.Readiness);
+                }
+            }
+
+            private void StartLivenessProbe(Probe probe, ReplicaState moveToOnSuccess = ReplicaState.Healthy)
+            {
+                // currently only HTTP is available
+                if (probe.Http is null)
+                {
+                    _logger.LogWarning("cannot start probing replica {name} because probe configuration is not set", _replica.Name);
```

Comment:
```suggestion
                    _logger.LogWarning("Cannot start probing replica {name} because probe configuration is not set", _replica.Name);
```

Reply:
And for the rest of the file.